core/arch/arm/mm/tee_mmu.c

```diff
@@ -496,245 +496,240 @@ static TEE_Result param_mem_to_user_va(struct user_ta_ctx *utc,
+	return alloc_pgt(utc);
 }
 
 TEE_Result tee_mmu_add_rwmem(struct user_ta_ctx *utc, struct mobj *mobj,
-			     int pgdir_offset, vaddr_t *va)
+			     vaddr_t *va)
 {
-	struct tee_ta_region *reg = NULL;
-	struct tee_ta_region *last_reg;
-	vaddr_t v;
-	vaddr_t end_v;
-	size_t n;
-
-	assert(pgdir_offset < CORE_MMU_PGDIR_SIZE);
-
-	/*
-	 * Avoid the corner case when no regions are assigned, currently
-	 * stack and code areas are always assigned before we end up here.
-	 */
-	if (!utc->mmu->regions[0].size)
-		return TEE_ERROR_GENERIC;
-
-	for (n = 1; n < ARRAY_SIZE(utc->mmu->regions); n++) {
-		if (!reg && utc->mmu->regions[n].size)
-			continue;
-		last_reg = utc->mmu->regions + n;
-
-		if (!reg) {
-			reg = last_reg;
-			v = ROUNDUP((reg - 1)->va + (reg - 1)->size,
-				    SMALL_PAGE_SIZE);
-#ifndef CFG_WITH_LPAE
-			/*
-			 * Non-LPAE mappings can't mix secure and
-			 * non-secure in a single pgdir.
-			 */
-			if (mobj_is_secure((reg - 1)->mobj) !=
-			    mobj_is_secure(mobj))
-				v = ROUNDUP(v, CORE_MMU_PGDIR_SIZE);
-#endif
+	TEE_Result res;
+	struct vm_region *reg = calloc(1, sizeof(*reg));
 
-			/*
-			 * If mobj needs to span several page directories
-			 * the offset into the first pgdir need to match
-			 * the supplied offset or some area used by the
-			 * pager may not fit into a single pgdir.
-			 */
-			if (pgdir_offset >= 0 &&
-			    mobj->size > CORE_MMU_PGDIR_SIZE) {
-				if ((v & CORE_MMU_PGDIR_MASK) <
-				    (size_t)pgdir_offset)
-					v = ROUNDDOWN(v, CORE_MMU_PGDIR_SIZE);
-				else
-					v = ROUNDUP(v, CORE_MMU_PGDIR_SIZE);
-				v += pgdir_offset;
-			}
-			end_v = ROUNDUP(v + mobj->size, SMALL_PAGE_SIZE);
-			continue;
-		}
+	if (!reg)
+		return TEE_ERROR_OUT_OF_MEMORY;
 
-		if (!last_reg->size)
-			continue;
-		/*
-		 * There's one registered region after our selected spot,
-		 * check if we can still fit or if we need a later spot.
-		 */
-		if (end_v > last_reg->va) {
-			reg = NULL;
-			continue;
-		}
-#ifndef CFG_WITH_LPAE
-		if (mobj_is_secure(mobj) != mobj_is_secure(last_reg->mobj) &&
-		    end_v > ROUNDDOWN(last_reg->va, CORE_MMU_PGDIR_SIZE))
-			reg = NULL;
-#endif
+	reg->mobj = mobj;
+	reg->offset = 0;
+	reg->va = 0;
+	reg->size = ROUNDUP(mobj->size, SMALL_PAGE_SIZE);
+	if (mobj_is_secure(mobj))
+		reg->attr = TEE_MATTR_SECURE;
+	else
+		reg->attr = 0;
+
+	res = umap_add_region(utc->vm_info, reg);
+	if (res) {
+		free(reg);
+		return res;
 	}
 
-	if (reg) {
-		TEE_Result res;
-
-		end_v = MAX(end_v, last_reg->va + last_reg->size);
-		res = alloc_pgt(utc, utc->mmu->ta_private_vmem_start, end_v);
-		if (res != TEE_SUCCESS)
-			return res;
-
-		*va = v;
-		reg->va = v;
-		reg->mobj = mobj;
-		reg->offset = 0;
-		reg->size = ROUNDUP(mobj->size, SMALL_PAGE_SIZE);
-		if (mobj_is_secure(mobj))
-			reg->attr = TEE_MATTR_SECURE;
-		else
-			reg->attr = 0;
-		return TEE_SUCCESS;
+	res = alloc_pgt(utc);
+	if (res) {
+		TAILQ_REMOVE(&utc->vm_info->regions, reg, link);
+		free(reg);
```

Comment:
minor: `TAILQ_REMOVE(&vm_info->regions, reg, link); free(reg);` is used at several places. Could deserve to be provided by a function as `umap_remove_region(&vm_info, reg);` ?

Reply:
OK, will fix